update collection_finder with warning for bad role names

Noel-Voelin opened the pull request:
add function to validate role names
SUMMARY

I added a little function that is validating the role_name according to the criteria documented in https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_structure.html#roles-directory

This is my first contribution so pelase feel free to come up with plenty of corrections and/or let me know what I have not considered.

Fixes #75023
ISSUE TYPE

    Bugfix Pull Request

COMPONENT NAME

update of RoleDefinition error handling
ADDITIONAL INFORMATION

Normally, I would solve this issue using a regular expression in python. However, I did not want to add the re module. If this would have been the way to go, let me know and I apply the changes accordingly.

This PR is a cleaned up version of: https://github.com/ansible/ansible/pull/75263

Noel-Voelin committed:
update collection_finder with warning for bad role names

apply validation for role names

Noel-Voelin commented:
I assume a unit test for this warning is needed too. I will look up on how to properly test warnings as I have never done that before.

bcoca commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -739,6 +740,9 @@ def __init__(self, collection_name, subdirs, resource, ref_type):
 
         if self.ref_type == u'role':
+            if not self.is_valid_role_name(resource):
+                warnings.warn("Role names are now limited to contain only lowercase alphanumeric characters, "
```
use display.warning instead (has internal dedupe to avoid warning spam)

Noel-Voelin committed:
use ansible.utils.display instead of warnings

exchange python warnings module with ansible.utils.display for internal dedupe handling

samdoran commented:
This is failing CI for valid reasons. Bigger picture, I don't think collection loader is the proper place to detect and report this. @nitzmahone may have a better idea.

s-hertel commented:
Roles from collections are loaded here https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/definition.py#L158-L159

It's hard to identify if an invalid role is actually part of a collection since `AnsibleCollectionRef.is_valid_fqcr(role_name)` will be False for a role name that isn't a valid Python identifier. We could try to make a guess at the end of the method if no roles have been found, maybe:

```diff
diff --git a/lib/ansible/playbook/role/definition.py b/lib/ansible/playbook/role/definition.py
index 20d69ebd6b..913aa49b23 100644
--- a/lib/ansible/playbook/role/definition.py
+++ b/lib/ansible/playbook/role/definition.py
@@ -200,7 +200,15 @@ class RoleDefinition(Base, Conditional, Taggable, CollectionSearch):
             return (role_name, role_path)
 
         searches = (self._collection_list or []) + role_search_paths
-        raise AnsibleError("the role '%s' was not found in %s" % (role_name, ":".join(searches)), obj=self._ds)
+        if self._collection_list or (len(role_name.split('.')) == 3 and not AnsibleCollectionRef.is_valid_fqcr(role_name)):
+            # Might be in a collection and not meet the naming requirements
+            hint = (
+                '\n\nHint: it looks like this might be a role in a collection. '
+                'Roles in collections need to be valid Python identifiers (for example, may not contain "-").'
+            )
+        else:
+            hint = ''
+        raise AnsibleError("the role '%s' was not found in %s%s" % (role_name, ":".join(searches), hint), obj=self._ds)
```

But that hint could be unhelpful in some cases, such as if you have a non-collections role named something like `contains.many.periods-and-invalid-chars`.

Noel-Voelin commented:
@s-hertel Since this 3 dot lookup as you mentioned runns the risk of raising warnings for non collection roles, I would suggest one of the following approaches:

1.) Use the levels (number of dots) for identification and if the role is considered to be within a collection, show an arbitrary warning that implies that the naming pattern could be violated. This message would have to be written in a way that it's clear that if the role is not in a collection this naming convention can be dismissed. 

This would provide neccesary information to people facing this issue and help resolving the problem more easily. However, as you mentioned already, the hint would be not helpful or perhaps even confusing for cases like contains.many.periods-and-invalid-chars.

2.) Alternatively to 1. In order to have a predictable and consistent solution, the message could be rewritten in a slightly more generic way to inform about the existance of this naming pattern even when the current role is not in a collection. This would be the technically least invasive way, but still achieve the goal.

Something like: AnsibleError("the role '%s' was not found in %s or the role is part of a collection and violates the collection specific naming pattern." % (role_name, ":".join(searches)), obj=self._ds)

3.) As a bit of an over the top approach, I assume the collection path could be used to to do a lookup for a galaxy.yml file. If it's present on the collections root level, we have confirmation that we're dealing with a role within a collection.

Not sure if 3. is even feasible.


If none of these seem a valid option, perhaps the issue is not sincere enough to follow up on it furteher and it could be closed with a little addition to the documentation which explains the error behavior in case the naming pattern is not respected. Thanks for all the help I have received so far! This is my first contribution to the Ansible project and I appreciate the support I am getting.

s-hertel commented:
> 1.)

There isn't a particular number of dots that indicates whether or not it's a role in a collection. If it's more than or equal to 3 or if the collections list is set, it's possible it could be part of a role. In a collection, '.'  in a role name indicates a directory (it's a python package), whereas for a standalone role it's just part of the directory name.

> 2.)

Just making the warning more inclusive of various the possibilities seems like a good idea.

> 3.)

A `galaxy.yml` isn't a requirement for a usable role. I don't think the collection_loader is the right place to do something like that either.

bcoca commented:
code lgtm, can we get a changelog and a test? also needs to pass tests (test has it's own collection loader that might need update), rerunning them in any case

/azp run

bcoca commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -739,6 +742,9 @@ def __init__(self, collection_name, subdirs, resource, ref_type):
         if self.ref_type == u'role':
+            if not self.is_valid_role_name(resource):
+                display.warning("Role names are now limited to contain only lowercase alphanumeric characters, "
+                                "plus _ and start with an alpha character. Please check %s" % resource)
```
i would rephrase a bit, 'roles names in collections are limited' .. standalone roles are not and collections always had this limitation.

bcoca replied:
probably should just be an error as we won't be able to find the role and later on we'll fail with 'role not found'

jborean93 commented:
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.
Our reasons for this are:

   * While we aren't against the idea of adding a warning to help users the current location is not desirable, we should either;
       * Add it after it has been installed and not part of the collection ref finder, or
       * Add it as a warning when it tries to load the role with a name like this
   * It would also be nice if this was done for any collection reference, i.e. any plugin and not just a role
   * Due to the age of the PR we are going to close this one, we are happy to review a new PR with a change based on the above if you wish to pursue this

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!
